readthedocs/proxito/README.rst

```diff
@@ -87,9 +88,9 @@ Writing a custom URL pattern
 The simplest way to write a custom URL pattern is to start from the default ones,
 and change them as needed.
 
-- Multi version project: ``^/{language}(/({version}(/{filename})?)?)?$``
-- Single version project: ``^/{filename}$``
-- Subproject: ``^/projects/{subproject}(/{filename})?$``
+- Multi version project: ``/{language}(/({version}(/{filename})?)?)?``
```

Comment:
I still don't quite understand why all the `?`'s are needed at the end here? It seems like we really want to be able to just say `/{language}/{version}/{filename}` right? Is it just to support slash redirects or something?

Reply:
> I think this is the reason? But it seems like we're adding a lot of complexity in the regex, and I don't quite understand what the value is that we're getting out of it?

We need partial matches to be able to have the correct 404 on a given component, and for redirects, and for when a path doesn't end with `/`.

> Can we post-process /{language}/{version}/{filename} to wrap it in a smart way? to get the regex that's above?

Maybe, it may get tricky when we have things between components `/{language}/something/{version}`, one of the reasons I wanted to avoid using regexes... I can try to see if I can get an algorithm that works reliable adding the groups and `?` where needed.

Reply:
`/{language}/something/{version}` feels like a pretty odd use for this. 

> We need partial matches to be able to have the correct 404 on a given component, and for redirects, and for when a path doesn't end with /.

I still don't quite understand why that is? Can we not just catch the URL in some kind of `.*` URL handler if the URLConf doesn't match, and still parse it in some smart way? 

I don't want to over-engineer this, so I'm probably OK shipping with what you've built. But I just forsee having issues with these regex's not being properly constructed because of how complex they are to get right..

Reply:
Since we are using `re.match`, we need one regex that allows us to match

 - /en
 - /en/
 - /en/latest
 - /en/latest/
 - /en/latest/file/name/

And the one for subprojects to match

- /projects/subproject
- /projects/subproject/
- /projects/subproject/file/name/

These are partial matches that we need to archive with just one regex, so the need for groups.

Reply:
I think I should be able to split each component on `/` or on the end of each `}`, and build a stack that adds the groups with `?`. Going to give it a try.

Reply:
Implemented this in https://github.com/readthedocs/readthedocs.org/pull/10156/commits/1ee0bd29cf0180c44d64a7cb62e851b24ef215fd, the algorithm to generate the groups is short, but it may look complex. URL patterns are now simple strings with replacement fields, no more defining of capture groups manually or escaping regex characters, but the complexity was moved to another place, so.

Let me know what you think, if we are good, I can update the proxito docs to match this new implementation, and add some more tests.

Reply:
In another comment I suggested to review what Django is doing to support "path converters" (https://docs.djangoproject.com/en/4.2/topics/http/urls/#path-converters) and, even, try to reuse what they have built already since they provide a way to expand the the converters by creating custom ones.

I think it's a good usage of our time to explore that idea hoping we can reuse their code and simplify ours a lot.

Reply:
>>  We need partial matches to be able to have the correct 404 on a given component, and for redirects, and for when a path doesn't end with /.
> I still don't quite understand why that is? 

This is still unclear to me. What do you mean by "be able to have the correct 404 on a given component"?